src/app/auth/email.ts

```diff
@@ -0,0 +1,79 @@
+import { EmailAlreadyExistsError } from "@domain/auth/errors"
+import { AuthWithEmailPasswordlessService } from "@services/kratos"
+import { baseLogger } from "@services/logger"
+import { UsersRepository } from "@services/mongoose"
+
+export const addEmailToIdentity = async ({
+  email,
+  userId,
+}: {
+  email: EmailAddress
+  userId: UserId
+}): Promise<AddEmailToIdentityResult | KratosError> => {
+  const authServiceEmail = AuthWithEmailPasswordlessService()
+
+  const hasEmail = await authServiceEmail.hasEmail({ kratosUserId: userId })
+  if (hasEmail instanceof Error) return hasEmail
+  if (hasEmail) return new EmailAlreadyExistsError()
```

Comment:
This has issues with idempotency.
If the `authServiceEmail.sendEmailWithCode ` call fails then the user entity hasn't been persisted yet. The end user might try again and hit the early return even though the `user` wasn't updated.

Reply:
I'm not sure to understand the concern? if `hasEmail === true` we don't want to add another email.

Reply:
You are updating 2 places - auth service + user entity. I'm just talking about the potential state / inconsistency that can arise when auth service update succeeds but user entity update fails. IE a likely outcome in a distributed system.

Reply:
how would you address this issue here?